### PR TITLE
 Add new HSM APIs to specify SAN entries when creating certificates

### PR DESCRIPTION
**Pull request by mrohera**
As part of the ID certificate capability within the security daemon, the HSM API has been extended to support adding Subject Alternate Names (SAN) entries to the certificate.

### CODE REVIEW

File: edgelet/hsm-sys/azure-iot-hsm-c/src/certificate_info.c [hunk 1]
```diff
@@ -602,10 +602,13 @@ void certificate_info_destroy(CERT_INFO_HANDLE handle)
     if (cert_info != NULL)
     {
         free(cert_info->first_certificate);
+        cert_info->first_certificate = NULL;
         free(cert_info->certificate_pem);
+        cert_info->certificate_pem = NULL;
         if (cert_info->private_key != NULL)
         {
             free(cert_info->private_key);
+            cert_info->private_key = NULL;
```
Comment (avranju) [inline edgelet/hsm-sys/azure-iot-hsm-c/src/certificate_info.c]: Curious, aren't these `NULL` assignments redundant given that `cert_info` itself is going to be freed on line 613?

File: edgelet/hsm-sys/azure-iot-hsm-c/src/hsm_certificate_props.c [hunk 2]
```diff
@@ -585,4 +612,99 @@ const char* get_alias(CERT_PROPS_HANDLE handle)
...
+    else
+    {
+        size_t i;
+        size_t list_size = num_san_entries * sizeof(char*);
+        destroy_san_entries(handle);
+        if (((handle->san_list = (char **)malloc(list_size)) == NULL) ||
+            ((handle->san_list_ro = (char const**)malloc(list_size)) == NULL))
+        {
+            LogError("Could not allocate memory for SAN list");
+            result = __LINE__;
```
Comment (avranju) [inline edgelet/hsm-sys/azure-iot-hsm-c/src/hsm_certificate_props.c]: What if the `san_list` allocation succeeds but the `san_list_ro` allocation fails? Will we leak `san_list`?

File: edgelet/hsm-sys/azure-iot-hsm-c/src/certificate_info.c (hunk 1 above)
Comment (mrohera) [reply edgelet/hsm-sys/azure-iot-hsm-c/src/certificate_info.c]: There is a chance that a client could use a freed handle which might have these stale entries. By zeroing out, we eliminate some aspect of "undefined behavior" and the failure will be a consistent NULL ptr deref.

File: edgelet/hsm-sys/azure-iot-hsm-c/src/hsm_certificate_props.c (hunk 2 above)
Comment (mrohera) [reply edgelet/hsm-sys/azure-iot-hsm-c/src/hsm_certificate_props.c]: Yes you are right, this is a possibility. I will fix this.

File: edgelet/hsm-sys/azure-iot-hsm-c/inc/hsm_certificate_props.h [hunk 3]
```diff
@@ -231,6 +234,33 @@ extern int set_alias(CERT_PROPS_HANDLE handle, const char* alias);
...
+);
+
+/**
+* @brief                Gets the alias type
+*
+* @param handle         The CERT_PROPS_HANDLE that was created by the cert_properties_create call
+* @param num_entries    The number of entries in the list will be returned
+*
+* @return               A pointer to a list of NULL terminated strings containing the SAN entries,
+*                       NULL otherwise.
```
[the remark this answers is not in the document]
Comment (mrohera) [reply edgelet/hsm-sys/azure-iot-hsm-c/inc/hsm_certificate_props.h]: Yes, I have updated the documentation to clarify.

File: edgelet/hsm-sys/azure-iot-hsm-c/src/hsm_certificate_props.c (hunk 2 above)
Comment (mrohera) [reply edgelet/hsm-sys/azure-iot-hsm-c/src/hsm_certificate_props.c]: Fixed and added negative test case as well to ensure coverage,

File: edgelet/hsm-sys/azure-iot-hsm-c/inc/hsm_certificate_props.h [hunk 4]
```diff
@@ -231,6 +234,33 @@ extern int set_alias(CERT_PROPS_HANDLE handle, const char* alias);
...
+*/
+extern int set_san_entries
+(
+    CERT_PROPS_HANDLE handle,
+    const char* san_list[],
+    size_t num_san_entries
+);
+
+/**
+* @brief                Gets the alias type
```
[the remark this answers is not in the document]
Comment (mrohera) [reply edgelet/hsm-sys/azure-iot-hsm-c/inc/hsm_certificate_props.h]: Fixed.

File: edgelet/hsm-sys/azure-iot-hsm-c/inc/hsm_certificate_props.h [hunk 5]
```diff
@@ -231,6 +234,33 @@ extern int set_alias(CERT_PROPS_HANDLE handle, const char* alias);
...
 extern const char* get_alias(CERT_PROPS_HANDLE handle);
 
+/**
+* @brief                  Sets the certificate subject alternate names
+*
+* @param handle           The CERT_PROPS_HANDLE that was created by the cert_properties_create call
+* @param san_list         A pointer to a list of string containing san entries
+* @param num_entries      The number of entries in the list
+*
+* @return                 On success 0 on.  Non-zero on failure.
```
[the remark this answers is not in the document]
Comment (mrohera) [reply edgelet/hsm-sys/azure-iot-hsm-c/inc/hsm_certificate_props.h]: Fixed,

**Pull request merged**